File: go/maximum_subarray.go
package leetcode

func maxSubArray(nums []int) int {
	res := nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i-1] > 0 {
			nums[i] += nums[i-1]
		}
		if nums[i] > res {
			res = nums[i]
		}
	}
	return res
}

// Using divide and conquer approach
func maxSubArray2(nums []int) int {
	return dAndd(0, len(nums)-1, nums)
}

func dAndd(l, r int, nums []int) int {
	if l == r {
		return nums[l]
	}

	m := l + (r-l)/2
	left := dAndd(l, m, nums)
	right := dAndd(m+1, r, nums)

	lMax, curr := nums[m], nums[m]
	for i := m - 1; i >= l; i-- {
		curr += nums[i]
		if curr > lMax {
			lMax = curr
		}
	}

	rMax, curr := 0, 0
	for i := m + 1; i <= r; i++ {
		curr += nums[i]
		if curr > rMax {
			rMax = curr
		}
	}

	if left >= right && left >= lMax+rMax {
		return left
	} else if right >= left && right >= lMax+rMax {
		return right
	} else {
		return lMax + rMax
	}
}

File: python/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res = nums[0]
        for i in range(1, len(nums)):
            if nums[i-1] > 0:
                nums[i] += nums[i-1]
            res = max(res, nums[i])
        return res

# using divide and conquer approach
class Solution:
    def maxSubArray2(self, nums: List[int]) -> int:
        def d_and_d(l, r, nums):
            if l == r: return nums[l]
            
            mid = l + (r - l) // 2
            left = d_and_d(l, mid, nums)
            right = d_and_d(mid+1, r, nums)
            
            l_max = curr = nums[mid]
            for i in range(mid-1, l-1, -1):
                curr += nums[i]
                l_max = max(l_max, curr)
            
            r_max = curr = 0
            for i in range(mid+1, r+1, 1):
                curr += nums[i]
                r_max = max(r_max, curr)
            
            return max(left, right, l_max+r_max)
        
        return d_and_d(0, len(nums)-1, nums)